### PR TITLE
Developer continuous plots

### PR DESCRIPTION
**Pull request by mpielies**
Pull request to merge developer_continuous_plots with developer_continuous

### CODE REVIEW

File: src/move/tasks/identify_associations.py [hunk 1]
```diff
@@ -20,12 +20,12 @@
...
 from move.data import io
 from move.data.dataloaders import MOVEDataset, make_dataloader
-from move.data.perturbations import perturb_categorical_data
+from move.data.perturbations import perturb_categorical_data, perturb_continuous_data_extended
 from move.data.preprocessing import one_hot_encode_single
 from move.models.vae import VAE
 
 TaskType = Literal["bayes", "ttest"]
-
+ContinuousTargetValue = ["minimum","maximum","plus_std","minus_std"]
```
[the remark this answers is not in the document]
Comment (ri-heme) [reply src/move/tasks/identify_associations.py:28]: Also, I configured isort in the [`pyproject.toml`](https://github.com/RasmussenLab/MOVE/blob/main/pyproject.toml) to keep it consistent with black.

File: src/move/tasks/encode_data.py [hunk 2]
```diff
@@ -42,8 +45,19 @@ def encode_data(config: DataConfig):
         logger.info(f"Encoding '{dataset_name}'")
         filepath = raw_data_path / f"{dataset_name}.tsv"
         names, values = io.read_tsv(filepath, sample_names)
+
+        # Plotting the value distribution for all continuous datasets before preprocessing:
+        fig = plot_value_distributions(values)
```
Comment (ri-heme) [inline src/move/tasks/encode_data.py:50]: Great idea to include before/after plots! 😄

File: src/move/tasks/identify_associations.py [hunk 3]
```diff
@@ -48,15 +48,24 @@ def _validate_task_config(
...
-    """Trains multiple models to identify associations between the dataset
-    of interest and the continuous datasets."""
+    """Leads to the execution of the appropiate association 
+    identification functions."""
+
+    task_config = cast(IdentifyAssociationsConfig, config.task)
+    if task_config.target_value in ContinuousTargetValue:
+        identify_associations_continuous(config)
+    else:
+        identify_associations_categorical(config)
```
Comment (ri-heme) [inline src/move/tasks/identify_associations.py:59]: Looks like there's not many differences between the approach functions (only the mask). So I think, it would be better to do something like:

```python
# pseudo-code
def identify_associations(config):
    if continuous_target:
        dataloaders, mask, etc = prepare_for_continuous_perturbation(config)
    else:
        dataloaders, mask, etc = prepare_for_categorical_perturbation(config)
    def _bayes_approach(config):
        # do bayes
    def _ttest_approach(config):
        # do ttest
    if bayes:
        results = _bayes_approach(config)
    elif ttest:
        results = _ttest_approach(config)
    save_results(results)
```

This makes the code easier to maintain. If there's something we need to change about the ttest/bayes, we only have to change it once in the code, and not fear we forgot to duplicate the fix in the other function.

**Review by ri-heme — COMMENTED** [review]
Minor comments (mostly formatting) and some suggestions, but great work! 😃 

Other general suggestion: give a descriptive name to your commits (optionally add a description if you need more explanation). That way, it's easier to look at the code history and track down when bugs were introduced or things stopped working.

Also, if you like, use the [emoji system](https://gitmoji.dev/) to tag your commits (makes the commit history more readable and structured).

File: src/move/data/perturbations.py [hunk 4]
```diff
@@ -84,7 +87,7 @@ def perturb_continuous_data(
     assert baseline_dataset.con_shapes is not None
     assert baseline_dataset.con_all is not None
 
-    target_idx = con_dataset_names.index(target_dataset_name)
+    target_idx = con_dataset_names.index(target_dataset_name)
```
[the remark this answers is not in the document]
Comment (mpielies) [reply src/move/data/perturbations.py:90]: Applied sort and black to all files

File: src/move/data/perturbations.py [hunk 5]
```diff
@@ -7,6 +7,9 @@
 from torch.utils.data import DataLoader
 
 from move.data.dataloaders import MOVEDataset
+from move.data.preprocessing import feature_stats
+from move.core.typing import PathLike
```
[the remark this answers is not in the document]
Comment (mpielies) [reply src/move/data/perturbations.py:11]: done! I applied isort and black to all files

File: src/move/data/perturbations.py [hunk 6]
```diff
@@ -109,3 +112,86 @@ def perturb_continuous_data(
...
+            batch_size=baseline_dataloader.batch_size,
+        )
+        dataloaders.append(perturbed_dataloader)
+
+
+    print(np.array(perturbations_list).transpose())
+
+    # Plot the perturbations for all features, collapsed in one plot:
+    fig = plot_value_distributions(np.array(perturbations_list).transpose())
+    fig_path = str(output_subpath / "Perturbation_distribution_{}.png".format(target_dataset_name))
```
[the remark this answers is not in the document]
Comment (mpielies) [reply src/move/data/perturbations.py:194]: Done

File: src/move/tasks/identify_associations.py [hunk 7]
```diff
@@ -302,3 +311,252 @@ def _ttest_approach(
...
+    def _bayes_approach(
+        task_config: IdentifyAssociationsBayesConfig,
+    ) -> tuple[IntArray, FloatArray]:
+        assert task_config.model is not None
+        # Train models
+        logger.info("Training models")
+        mean_diff = np.zeros((num_perturbed, num_samples, num_continuous))
+        normalizer = 1 / task_config.num_refits
+        for j in range(task_config.num_refits):
+            print("{}/{}".format(j+1,task_config.num_refits))
```
[the remark this answers is not in the document]
Comment (mpielies) [reply src/move/tasks/identify_associations.py:387]: f-string formatted and turned into logger.info

File: src/move/visualization/dataset_distributions.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        colormap:
+            Name of colormap to apply to the colorbar.
+
+    Returns:
+        Figure
+    """
+    width: float = max(matplotlib.rcParams["figure.figsize"])
+    vmin, vmax = np.nanmin(feature_values),  np.nanmax(feature_values)
+    with style_settings(style):
+        fig = plt.figure(layout='constrained')
```
[the remark this answers is not in the document]
Comment (mpielies) [reply src/move/visualization/dataset_distributions.py:38]: I applied black to all files.

File: src/move/visualization/dataset_distributions.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+            Values of the features, a 2D array (`num_samples` x `num_features`).
+        style:
+            Name of style to apply to the plot.
+        colormap:
+            Name of colormap to apply to the colorbar.
+
+    Returns:
+        Figure
+    """
+    width: float = max(matplotlib.rcParams["figure.figsize"])
```
[the remark this answers is not in the document]
Comment (mpielies) [reply src/move/visualization/dataset_distributions.py:35]: var removed

File: src/move/conf/schema.py [hunk 10]
```diff
@@ -216,6 +216,7 @@ class IdentifyAssociationsConfig(TaskConfig):
     num_refits: int = MISSING
     sig_threshold: float = 0.05
     save_refits: bool = False
+    perturbation_continuous: str = MISSING
```
[the remark this answers is not in the document]
Comment (mpielies) [reply src/move/conf/schema.py:219]: Yes, I removed it but I had some issues with git stash, which probably kept an older version of the file.

File: tutorial/config/task/random_small__id_assoc_bayes.yaml [hunk 11]
```diff
@@ -4,10 +4,10 @@ defaults:
 
 batch_size: 10
 
-num_refits: 40
+num_refits: 40 # default is 40
 
-target_dataset: random.small.drugs
-target_value: 1
+target_dataset: random.small.metagenomics
+target_value: maximum
```
[the remark this answers is not in the document]
Comment (mpielies) [reply tutorial/config/task/random_small__id_assoc_bayes.yaml:10]: created the new yaml files

File: src/move/data/perturbations.py [hunk 12]
```diff
@@ -109,3 +112,86 @@ def perturb_continuous_data(
...
+            shuffle=False,
+            batch_size=baseline_dataloader.batch_size,
+        )
+        dataloaders.append(perturbed_dataloader)
+
+
+    print(np.array(perturbations_list).transpose())
+
+    # Plot the perturbations for all features, collapsed in one plot:
+    fig = plot_value_distributions(np.array(perturbations_list).transpose())
```
[the remark this answers is not in the document]
Comment (mpielies) [reply src/move/data/perturbations.py:194]: Added the optional, but still output_subpath is defined inside the code downstream

File: src/move/visualization/dataset_distributions.py [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+from move.core.typing import FloatArray
+from move.visualization.style import (
+    DEFAULT_PLOT_STYLE,
+    style_settings,
+)
+
+def plot_value_distributions(
+    feature_values: FloatArray,
+    style: str = 'fast',
```
[the remark this answers is not in the document]
Comment (mpielies) [reply src/move/visualization/dataset_distributions.py]: Maybe we can do this together?

File: src/move/tasks/identify_associations.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mpielies) [reply src/move/tasks/identify_associations.py]: reformatted

**Comment by ri-heme** [comment]
I don't know why this got closed 😅

**Comment by ri-heme** [comment]
Closing this PR, as we will be merging another branch into the main branch
